torch/distributed/checkpoint/state_dict.py

```diff
@@ -563,6 +564,29 @@ def _init_optim_state(optim: torch.optim.Optimizer) -> None:
     optim.zero_grad(set_to_none=True)
 
 
+def _flatten_optim_state_dict(state_dict: OptimizerStateType) -> Dict[str, ValueType]:
+    def _raise_if_type_not_supported(v):
+        if not isinstance(v, (torch.Tensor, int, float)):
+            raise NotImplementedError(
+                "Flattening optimizer state_dict only supports "
+                "tensor, int, float states now. "
+                f"Type is {type(v)}."
+            )
+
+    ret: Dict[str, ValueType] = {}
+    for fqn, state in state_dict[STATE].items():
+        for k, v in state.items():
+            _raise_if_type_not_supported(v)
+            ret[f"{STATE}.{fqn}.{k}"] = v
+
+    for param_group in state_dict[PG]:
+        fqns = param_group.pop(PARAMS)
+        for fqn in fqns:
+            for k, v in param_group.items():
+                ret[f"{PG}.{fqn}.{k}"] = v
```

Comment:
it looks like this flattening would not preserve the original param-group groupings.  It is a true flattening.  Is that correct?

I'm not sure if that's good or bad- does that mean that we make the optimizer states generic enough that resharing would be possible?  (But I'm not sure how we'd reconstruct the original param_group groupings on load side.)

Reply:
ok reading the 'unflattening' side again it makes sense.  IIUC, because the user would have created the new optimizer based on the new model/parallelisms, it may or may not have the same param_groupings as before.  But since we normalized all of the individual states to fqn, as long as they have the same FQNs (or less FQNs) in their new model, they can find the states and then group them according to the new optimizer's groupings.

So this appears to support resharding?

Reply:
This supports resharding.